# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: gfx/wr/webrender/src/render_backend.rs
# Commit: 971767c6ed2b
# Full Hash: 971767c6ed2b2201e172105046ecc61942e03a7e
# Author: sotaro <sotaro.ikeda.g@gmail.com>
# Date: 2020-01-24 21:46:30
# Regressor Bug: 1602643
# File Overlap Count: 1
# Description:
#   Bug 1602643 - Disable WebRender compositor dinamically for async screenshot r=gw
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D59658
# ==============================================================================

diff -r 436df7dc06c6 -r 971767c6ed2b gfx/wr/webrender/src/render_backend.rs
--- a/gfx/wr/webrender/src/render_backend.rs	Fri Jan 24 01:04:09 2020 +0000
+++ b/gfx/wr/webrender/src/render_backend.rs	Thu Jan 23 23:52:05 2020 +0000
@@ -711,6 +711,7 @@
     resource_cache: ResourceCache,
 
     frame_config: FrameBuilderConfig,
+    default_compositor_kind: CompositorKind,
     documents: FastHashMap<DocumentId, Document>,
 
     notifier: Box<dyn RenderNotifier>,
@@ -755,6 +756,7 @@
             resource_cache,
             gpu_cache: GpuCache::new(),
             frame_config,
+            default_compositor_kind : frame_config.compositor_kind,
             documents: FastHashMap::default(),
             notifier,
             recorder,
@@ -1225,6 +1227,33 @@
                         self.resource_cache.clear(mask);
                         return RenderBackendStatus::Continue;
                     }
+                    DebugCommand::EnableNativeCompositor(enable) => {
+                        // Default CompositorKind should be Native
+                        if let CompositorKind::Draw { .. } = self.default_compositor_kind {
+                            unreachable!();
+                        }
+
+                        let compositor_kind = if enable {
+                            self.default_compositor_kind
+                        } else {
+                            CompositorKind::default()
+                        };
+
+                        for (_, doc) in &mut self.documents {
+                            doc.scene.config.compositor_kind = compositor_kind;
+                            doc.frame_is_valid = false;
+                        }
+
+                        self.frame_config.compositor_kind = compositor_kind;
+
+                        // Update config in SceneBuilder
+                        self.low_priority_scene_tx.send(SceneBuilderRequest::SetFrameBuilderConfig(
+                            self.frame_config.clone()
+                         )).unwrap();
+
+                        // We don't want to forward this message to the renderer.
+                        return RenderBackendStatus::Continue;
+                    }
                     DebugCommand::SimulateLongSceneBuild(time_ms) => {
                         self.scene_tx.send(SceneBuilderRequest::SimulateLongSceneBuild(time_ms)).unwrap();
                         return RenderBackendStatus::Continue;
@@ -1514,7 +1543,7 @@
         // external image with NativeTexture or when platform requested to composite frame.
         if invalidate_rendered_frame {
             doc.rendered_frame_is_valid = false;
-            if let CompositorKind::Draw { max_partial_present_rects } = self.frame_config.compositor_kind {
+            if let CompositorKind::Draw { max_partial_present_rects } = doc.scene.config.compositor_kind {
               // When partial present is enabled, we need to force redraw.
               if max_partial_present_rects > 0 {
                   let msg = ResultMsg::ForceRedraw;